README.md
## Introduction

This python script implements the [xkcd password spec](http://xkcd.com/936/).

## Install

This package requires the `pip` Python package manager for installation. [pip installation instructions](http://www.pip-installer.org/en/latest/installing.html).

Then:

```
pip install xkpa
```

## Usage

```
% xkpa -h
usage: xkpa.py [-h] [-n] [-d DICT_PATH] [-x] [-i] [-s SEPARATOR] [-l LENGTH]
               [-c COUNT] [-v]
               [w]

Generate an xkcd style password.

positional arguments:
  w             The number of words in the password. Defaults to 4.

optional arguments:
  -h, --help    show this help message and exit
  -n            Disable printing a newline at the end of the password. Good
                for piping to the clipboard.
  -d DICT_PATH  The dictionary file. Defaults to [PATH].
  -x            Disable excluding special characters and punctuation.
  -i            Enable showing password information (entropy, etc).
  -s SEPARATOR  Delimit words with a given character/string.
  -l LENGTH     The maximum word length. Words must be at or below this
                length.
  -c COUNT      Number of passwords to generate. Defaults to 1.
  -v            Display version information.

http://xkcd.com/936/
```

## Examples
**Generate a password, and put it in the clipboard (OS X)**

    xkpa -n | pbcopy

**Generate a password, copy to clipboard, and print info**

    xkpa -ni | pbcopy
    Info:
      Entropy: 61.292 bits
      Entropy per word: 15.323 bits
      Possible combinations given settings: 2.0 million trillion
      Average time to crack at 1.0 thousand tries per second: 44.7 million years
      Average time to crack at 70.0 thousand tries per second: 629.8 thousand years
      Average time to crack at 75.0 million tries per second: 580.8 years
      Average time to crack at 350.0 billion tries per second: 0.1 years

The password will get copied to the clipboard, but the info message will not (the info message is printed to stderr).

**Generate a password with info**

    xkpa -i
    wincing-peaceful-transplanted-nitrated
    
    Info:
      Entropy: 61.292 bits
      Entropy per word: 15.323 bits
      Possible combinations given settings: 2.0 million trillion
      Average time to crack at 1.0 thousand tries per second: 44.7 million years
      Average time to crack at 70.0 thousand tries per second: 629.8 thousand years
      Average time to crack at 75.0 million tries per second: 580.8 years
      Average time to crack at 350.0 billion tries per second: 0.1 years

**Generate a password with 5 words separated by '.'**

    xkpa -s '.' 5
    belies.annoyingly.birthing.ventilate.icon

**Generate a password with words no longer than 5 characters in length**

    xkpa -l 5
    cover-kings-yard-store

**Generate multiple passwords**

    xkpa -c 5
    1. excite-incriminating-wronged-veto
    2. ordinariest-reverberation-remain-subsidy
    3. noisiness-faraway-country-sides-straightens
    4. hardening-bribes-children-helpings
    5. calmes-touts-audition-cleaned

This is useful if someone is looking over your shoulder. By generating multiple passwords, someone looking over your shoulder will not know which is your real password. Each password is numbered, because your choice should be decided *before the passwords are generated* (if you generate 20 passwords, decided on a number between 1 and 20 beforehand). This prevents the urge to cherry pick passwords and decrease entropy.

## Cherry Picking Passwords

If the generated passwords contain too many esoteric words, *resist the urge to cherry pick passwords with easier to remember words.* This will decrease the entropy of the password in an unpredictable way. The proper way to generate passwords with more familiar words is to use the `-l` option to limit the length of each word. Although this will also decrease entropy, it will be done in such a way that the `-i` flag will still be able to calculate the entropy of the password. When using the `-l` flag, be sure to use it in conjunction with `-i` to ensure the entropy of the password is suitable for your needs.

So, use `xkpa -i -l NUM` until you've found a setting with the desired entropy. *Then generate one password and use it.* Don't keep generating a password until you've found one you "like."

Do not use the `-c` flag to generate several passwords at once, and cherry pick. This option is meant for use in public places where someone may be looking over your shoulder. The proper way to use this option is to think of a number between 1 and 100, and then generate 100 passwords with `xkpa -c 100`. Then choose the password which corresponds to the number you've already picked. This way, you're not cherry picking, and if someone glances over your shoulder, they won't know which of the 100 passwords is your real password.

## Entropy

How much entropy you need really depends on what you're trying to protect against. State of the art password crackers can crack Windows password files at a rate of [300+ billion attempts per second](http://passwords12.at.ifi.uio.no/Jeremi_Gosney_Password_Cracking_HPC_Passwords12.pdf) [PDF] and this rate will only increase in the future. On the other hand, brute forcing a remote sever will take place at a much lower rate, and secure servers should have throttling mechanisms in place making this sort of attack obsolete. So, if you're worried that someone will try to reverse your hashed password, and you think your passwords are being hashed with a weak algorithm (e.g., MD5 or NTLM), then aim for high entropy. If you're generating a password for a remote server, and that server has throttling mechanisms in place, or your passwords are being hashed with a more secure algorithm (e.g., bcrypt), a lower entropy may be acceptable.

The information on time-to-crack given by the `-i` flag was specifically chosen because these values represent the current state of the art. Weak hashes, like NTLM and MD5 can be attempted at 350 billion and 180 billion tries per second, respectively. bcrypt, a more secure hash, slows this same attack down to 70 thousand attempts per second (75 million/sec for md5crypt). Attacks against remote servers probably happen somewhere within an order of magnitude of 1,000 attempts per second against a poorly secured server, and would be immediately throttled against a secure server. [CloudCracker](https://www.cloudcracker.com/) boasts a rate of 300,000,000 words in 20 minute, or 41,000 words per second.

## Config File

There is no config file. Set default arguments by adding an alias to your `~/.bashrc`:

    alias pgen="xkpa -nil 5 5"

## Dictionary File
The dictionary file bundled with the script is from the `wamerican-small` package off Ubuntu. It strikes a nice balance between entropy and not having too many esoteric words. See README\_DICT.txt for the associated licenses and credits.

Modify the `DEFAULT_DICT` variable at the top of the script to use a different default.

## Randomness
This uses Python's [`random.SystemRandom`](https://docs.python.org/3/library/random.html#random.SystemRandom) as a source of entropy. This provides cryptographically secure randomness.

## License
The `xkpa.py` script is released under the MIT license:

```
Copyright (c) 2012 Alex Beal

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```

The dictionary file (`dict.txt`) is distributed under the license of its original authors located in the README\_DICT.txt file.


CHANGES.txt
v 0.1.4 (3/1/2017)
- Python 3 compatibility (Credit: Jason Alonso https://github.com/jbalonso)

v 0.1.3 (3/24/2013)
- Switched to PyCrypto's PRNG.
- Cleaned up info messages (chose random numbers for tries/sec).
- Simplified structure. Removed Dictionary class.

v 0.1.2 (12/8/2012)
- Adds more information in the "time to crack" info message. State of the art
  hash crackers can try 300+ billion hashes per sec. Message updated to reflect
  this.
- Adds advice on entropy to readme.

v 0.1.1 (8/7/2012)
- Removes low mem algorithm. Added complexity without really being useful.
- Adds -v for version information.
- Fixes -n bug when used in conjunction with -c.
- Adds CHANGES.txt file.


xkpass.py
#!/usr/bin/python

import random
import re
import math
import argparse

DEFAULT_DICT="/usr/share/dict/words"
WORDS = 4
BAD_CH_LIST = ['\'']

BAD_CHARS= "|".join(BAD_CH_LIST)

def createParser():
    parser = argparse.ArgumentParser(
            description='Generate an xkcd style password.',
            epilog='http://xkcd.com/936/')
    parser.add_argument('w',
            type=int,
            default=WORDS,
            nargs='?',
            help="The number of words to generate.")
    parser.add_argument('-d',
            type=open,
            default=DEFAULT_DICT,
            help="The dictionary file.")
    parser.add_argument('-x',
            action='store_false',
            default=True,
            help="Disable excluding special characters and punctuation.")
    parser.add_argument('-i',
            action='store_true',
            default=False,
            help="Enable showing password information (entropy, etc).")
    parser.add_argument('-s',
            default='-',
            help="Delimit words with a given character.")
    return parser

def loadDict(dict_file, exclude_char):
    dict_list = []
    for line in dict_file:
        word = line.strip()
        if exclude_char and re.search(BAD_CHARS, word):
            continue
        dict_list.append(word)
    return dict_list

def makePassword(dict_list, word_count, delimit):
    dict_len = len(dict_list)
    r = random.SystemRandom()
    pass_list = []
    for i in range(word_count):
        pass_list.append( dict_list[r.randint(0,dict_len-1)] )
    return delimit.join(pass_list)

def getPassInfo(dict_len, word_count):
    entropy = math.log(dict_len ** word_count, 2)
    return "\nInfo:\n  Entropy: %d\n  Entropy per word: %d" % (entropy, (entropy/word_count))

if __name__ == "__main__":
    # Parse arguments
    parser = createParser()
    args = vars( parser.parse_args() )
    word_count = args['w']
    dict_file = args['d']
    exclude_chars = args['x']
    info = args['i']
    delimit = args['s']

    #Read in dictionary
    dict_list = loadDict(dict_file, exclude_chars)

    # Select random words and create password
    print makePassword(dict_list, word_count, delimit)

    if info:
        print getPassInfo(len(dict_list), word_count)


experimental/dice_rolls.py
#!/usr/bin/python
import random
import operator

# This script rolls a bunch of dice, and then records which ones have the
# highest values. This is run for multiple trials. Ties are broken
# randomly. This experiment seems to indicate that each die has an equal
# chance of being in the top 3 (or 4 or 5, etc).

DICE = 10
RANGE = 1000
HIGHEST = 5
TRIALS = 100000
r = random.SystemRandom()

class die(object):
    def __init__(self, id, range):
        self.value = 0
        self._range = range
        self.id = id
        self.count = 0

    def roll(self):
        r.randint(1, self._range)

    def __cmp__(self, other):
        if self.value < other.value:
            return -1
        elif self.value > other.value:
            return 1
        else:
            return -1 if r.randint(0,1) == 0 else 1

    def __repr__(self):
        return "(id: %d, count: %d)" % (self.id, self.count)

dice_list = [ die(id, RANGE) for id in range(DICE) ]
for trial in range(TRIALS):
    for die in range(DICE):
        dice_list[die].roll()
    dice_list.sort()
    for die in range(HIGHEST):
        dice_list[die].count += 1
    dice_list = sorted(dice_list, key=lambda die:die.id)

print sorted(dice_list, key=lambda die : die.id)


experimental/rand_proto.py
#!/usr/bin/python -O

import random

DICT_PATH = "/usr/share/dict/words"
WORDS = 4
RANGE = 5000

def trim(d):
    word_list = d.items()
    word_list.sort()
    word_count = 0
    entry = 0
    while word_count < WORDS and entry < len(word_list):
        word_count += len( word_list[entry][1] )
        entry += 1
    remove = len(word_list) - entry
    if remove > 0:
        for e in range(0, remove):
            word_list.pop()
    return dict(word_list)

def grab(d, words):
    word_list = d.items()
    word_list.sort()
    print word_list
    count = 0
    selected_words = []
    while count < words:
        if len( word_list[0][1] ) == 0:
            del word_list[0]
        selected = word_list[0][1][r.randint(0, len(word_list[0][1])-1)]
        word_list[0][1].remove(selected)
        count += 1
        selected_words.append(selected)
    return selected_words

r = random.SystemRandom()
dict_file = file(DICT_PATH)
word_dict = {}
for line in dict_file:
    word = line.strip()
    word_value = r.randint(0, RANGE)
    if word_value in word_dict:
        word_dict[word_value].append(word)
    else:
        word_dict[word_value] = [word]
    word_dict = trim(word_dict)

print grab(word_dict, WORDS)


xkpa/xkpa.py
#!/usr/bin/python -O
 
from __future__ import print_function
import random
import math
import argparse
import os
import sys
from pkg_resources import resource_filename

PY2 = sys.version_info < (3,)
if PY2:
    range = xrange

# The dictionary bundled with the script. It must be in the same
# dir as the script.
DEFAULT_DICT =  resource_filename(__name__, 'dict.txt')
# Uncomment to use the system dictionary as the default.
#DEFAULT_DICT="/usr/share/dict/words"
WORDS = 4
BAD_CH_LIST = ['\'']
VERSION_NUMBER="0.1.4"

class WordValidator(object):
    """Validates words from the dictionary file, when the file is loaded
       into memory.
    """
    def __init__(self, flags):
        self._badChars = set(BAD_CH_LIST)
        self._maxLen = flags.l
        self._skipBad = flags.x

    def isValidWord(self, word):
        """Returns True if the word is valid. False otherwise.
        """
        if len(word) > self._maxLen or len(word) == 0:
            return False
        if self._skipBad and self._hasBadChar(word):
            return False
        return True

    def _hasBadChar(self, word):
        """Return true if word has a 'bad' char.
        """
        for char in reversed(word):
            if char in self._badChars:
                return True

class WordList(object):
    """List-like object that returns words from a dictionary file.
    """
    def __init__(self, flags):
        self._wordValidator = WordValidator(flags)
        self._loadDict(flags.d)

    def __getitem__(self, key):
        return self._dictList[key]

    def __len__(self):
        return len(self._dictList)

    def _loadDict(self, path):
        """Load every word from dict_file that passes the isValidWord test.
        """
        self._dictList = []
        with open(path) as dict_file:
            for line in dict_file:
                word = line.strip()
                if self._wordValidator.isValidWord(word):
                    self._dictList.append(word)

class RandomWord(object):
    """Iterator that returns on random word per iteration.
    """
    def __init__(self, flags):
        self._dictObj = WordList(flags)
        self._rng = random.SystemRandom()

    def __iter__(self):
        return self

    def __len__(self):
        return len(self._dictObj)

    def next(self):
        return self._rng.choice(self._dictObj)

    __next__ = next

class PasswordGen(object):
    """Iterator that returns one password per iteration.
    """
    def __init__(self, flags):
        self._randSource = RandomWord(flags)
        self._wordCount = flags.w
        self._delimiter = flags.s

    def __iter__(self):
        return self

    def next(self):
        rand_source_iter = iter(self._randSource)
        word_list = [rand_source_iter.next() for i in range(self._wordCount)]
        return self._delimiter.join(word_list)

    __next__ = next

    def getInfo(self):
        entropy, entr_per_word = self.calcEntropy()
        info =  "\nInfo:\n"
        info += "  Entropy: %0.3f bits\n" % entropy
        info += "  Entropy per word: %0.3f bits\n" % entr_per_word
        info += "  Possible combinations given settings: %s\n" % self.readableNum(self.calcPos())
        for triespersec in [1e3, 70e3, 75e6, 350e9]:
            info += self.makeYrsToCrackMsg(triespersec)
        return info

    def calcEntropy(self):
        entropy = math.log(self.calcPos(), 2)
        return entropy, entropy/self._wordCount

    def calcYrsToCrack(self, triesPerSec):
        try:
            return (self.calcPos()/(60*60*24*365.25*triesPerSec))/2
        except OverflowError:
            # If float calculation overflows, estimate using long.
            return (self.calcPos()/(60*60*24*365*int(triesPerSec)))/2

    def makeYrsToCrackMsg(self, triesPerSec):
        try:
            return "  Average time to crack at %s tries per second: %s years\n" % (self.readableNum(triesPerSec), self.readableNum(self.calcYrsToCrack(triesPerSec)))
        except TypeError:
            # If float conversion overflows, use long.
            return "  Average time to crack at %s tries per second: %s years\n" % (self.readableNum(triesPerSec), self.calcYrsToCrack(triesPerSec))

    def calcPos(self):
        return len(self._randSource) ** self._wordCount

    def readableNum(self, num, postfix=""):
        num_words = { 1000000000000: 'trillion ',
                      1000000000   : 'billion ',
                      1000000      : 'million ',
                      1000         : 'thousand ', }
        for div, word in num_words.items():
            if num/div >= 1:
                return self.readableNum(num/div, word + postfix)
        return "{:,.1f} {:s}".format(num, postfix).rstrip()


class Enumerator(object):
    """Accepts an iterator and prepends 1., 2., 3., etc to the beginning
       of each item returned from the generator.
    """
    def __init__(self, gen):
        self.gen = gen
        self.count = 0

    def __iter__(self):
        return self

    def next(self):
        self.count += 1
        return "%d. %s" % (self.count, self.gen.next())


def createParser():
    """Create the parser for the command line flagsi
    """
    # Default values for command line flags.
    flag_defaults = {
        'w': WORDS,
        'n':False,
        'd':DEFAULT_DICT,
        'x':True,
        'i':False,
        's':'-',
        'l':100,
        'c':1,
        'v':False,
        }
    parser = argparse.ArgumentParser(
            description='Generate an xkcd style password.',
            epilog='http://xkcd.com/936/')
    parser.add_argument('w',
            type=int,
            default=flag_defaults['w'],
            nargs='?',
            help="The number of words in the password. Defaults to 4.")
    parser.add_argument('-n',
            action='store_true',
            default=flag_defaults['n'],
            help="Disable printing a newline at the end of the password.\
                  Good for piping to the clipboard.")
    parser.add_argument('-d',
            default=flag_defaults['d'],
            metavar="DICT_PATH",
            help="The dictionary file. Defaults to %s." % DEFAULT_DICT)
    parser.add_argument('-x',
            action='store_false',
            default=flag_defaults['x'],
            help="Disable excluding special characters and punctuation.")
    parser.add_argument('-i',
            action='store_true',
            default=flag_defaults['i'],
            help="Enable showing password information (entropy, etc).")
    parser.add_argument('-s',
            default=flag_defaults['s'],
            metavar="SEPARATOR",
            help="Delimit words with a given character/string.")
    parser.add_argument('-l',
            default=flag_defaults['l'],
            type=int,
            metavar="LENGTH",
            help="The maximum word length. Words must be at or below this length.")
    parser.add_argument('-c',
            default=flag_defaults['c'],
            type=int,
            metavar="COUNT",
            help="Number of passwords to generate. Defaults to 1.")
    parser.add_argument('-v',
            action='store_true',
            default=flag_defaults['v'],
            help="Display version information.")
    return parser

def main():
    parser = createParser()
    flags = parser.parse_args()

    if flags.l < 1:
        sys.stderr.write("ERROR: Maxmimum word length is less than 1.\n")
        return
    if flags.c < 1:
        sys.stderr.write("ERROR: Password count is less than 1.\n")
        return

    if flags.v:
        print("Version: " + VERSION_NUMBER)
        return

    pGen = iter(PasswordGen(flags))
    # Print the password without adding a \n or space.
    password_count = int(flags.c)
    if password_count == 1:
        sys.stdout.write(pGen.next())
    elif password_count > 1:
        pGenEnumerate = iter(Enumerator(pGen))
        for i in range(password_count):
            sys.stdout.write(pGenEnumerate.next() + ("\n" if i != flags.c - 1 else ""))
    # Print newline if newline not disabled.
    if not flags.n:
        print("")

    # Make sure the password gets printed before info.
    sys.stdout.flush()
    if flags.i:
        sys.stderr.write(pGen.getInfo())

if __name__ == "__main__":
    main()
